fc/parsing/actions.py

```diff
@@ -1413,63 +1630,53 @@ def _handle_clamping(self):
-            units.
+    def _gather_state_variables(self, original_ordering):
+        """
+        Gather all variables annotated as states, and create a vector ordering. If required, update the ProtocolVariable
+        for the oxmeta:state_variable annotation.
+
+        :param original_ordering: An list containing the rdf identities of the variables as they originally appeared
+            (before any model manipulation).
         """
-        output_variables = set()
-        for output in self.outputs:
-            variables = get_variables_transitively(self.model, output.rdf_term)
-            if output.units is not None:
-                desired_units = self.units.get_unit(output.units)
-                for i, variable in enumerate(variables):
-                    variables[i] = self.model.convert_variable(
-                        variable, desired_units, DataDirectionFlow.OUTPUT)
-            output_variables.update(variables)
-        return output_variables
-
-    def _list_parameters(self):
-        """Populates the list of model parameters (inputs with a constant RHS)."""
-
-        for var in self.inputs:
+        # Gather state variables
+        states = get_variables_that_are_version_of(self.model, STATE_ANNOTATION)
+
+        # Create and store ordering, preserving original ordering as much as possible
+        order = []
+        todo = set([var.rdf_identity for var in states])
+        for rdf_identity in original_ordering:
             try:
-                variable = self.model.get_variable_by_ontology_term(var.rdf_term)
+                todo.remove(rdf_identity)
             except KeyError:
-                # Skip optional variables
                 continue
+            order.append(rdf_identity)
+        order.extend(todo)
+        self.vector_orderings[STATE_ANNOTATION] = {rdf_identity: i for i, rdf_identity in enumerate(order)}
 
-            # Parameters are inputs that aren't states, and have a constant RHS
-            eq = self.model.get_definition(variable)
-            if isinstance(eq.lhs, VariableDummy) and len(eq.rhs.atoms(VariableDummy)) == 0:
-                self.parameters.append(var)
+        # Set transitive variables for `state_variable` output, if present
+        if self.magic_pvar is not None:
+            self.magic_pvar.transitive_variables = set(states)
 
-    def _purge_unused_mathematics(self, output_variables):
-        """Remove model equations and variables not needed for generating desired outputs.
+    def _purge_unused_mathematics(self):
+        """Remove model equations and variables not needed for generating desired outputs."""
 
-        :param output_variables: the set of variables appearing in outputs, either directly or as part of a vector
-        """
+        # Output variables
+        required_variables = set()
+        for pvar in self.protocol_variables:
+            if pvar.output or pvar.output_category:
```

Comment:
It occurs to me that we should probably remove this line, since input variables should be considered required. It'd be a pretty bizarre protocol that defined an input which wasn't used in computing any outputs, but if someone did do so they'd be pretty confused as to why their input didn't exist!